pipeline.train accepts `DataSource`s

dcfidalgo opened the pull request:
In this PR the main change is that `Pipeline.train` and `Pipeline.explore`now also accept `DataSource`s instead of yaml files.

@frascuchon @dvsrepo Actually i would like to be a bit more radical and remove the option of passing on a yaml file path to these methods to keep the code simple and to "have only one obvious way to do stuff". Also, if you have a yaml file, you can always do:
```python
...
    training=DataSource.from_yaml(path),
...
```
What do you think?

dvsrepo commented:
Totally agree.

El lun., 25 may. 2020 20:02, David Fidalgo <notifications@github.com>
escribió:

> In this PR the main change is that Pipeline.train and Pipeline.explorenow
> also accept DataSources instead of yaml files.
>
> @frascuchon <https://github.com/frascuchon> @dvsrepo
> <https://github.com/dvsrepo> Actually i would like to be a bit more
> radical and remove the option of passing on a yaml file path to these
> methods to keep the code simple and to "have only one obvious way to do
> stuff". Also, if you have a yaml file, you can always do:
>
> ...
>     training=DataSource.from_yaml(path),
> ...
>
> What do you think?
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/recognai/biome-text/pull/224
> Commit Summary
>
>    - pipeline.train and pipeline.explore accept also `DataSource`s
>    instead of yaml files + remove attributes from `DataSource`
>
> File Changes
>
>    - *M* examples/1.text_classifier/text_classifier.py
>    <https://github.com/recognai/biome-text/pull/224/files#diff-9969e39ee661b1e8fba7374a9d8b17fb>
>    (10)
>    - *M* examples/2.document_classifier/document_classifier.py
>    <https://github.com/recognai/biome-text/pull/224/files#diff-26b59e6c196c0a2ab43a1d2986aeaf4e>
>    (4)
>    - *M* examples/3.email_classifier/email_classifier.py
>    <https://github.com/recognai/biome-text/pull/224/files#diff-1b9967f93ba6d4c30f13462a15198276>
>    (2)
>    - *M* src/biome/text/_configuration.py
>    <https://github.com/recognai/biome-text/pull/224/files#diff-268b91bf00efa76d2481b7564aa64331>
>    (20)
>    - *M* src/biome/text/_helpers.py
>    <https://github.com/recognai/biome-text/pull/224/files#diff-b81d5d578d688af7e6c79b6b33b1293e>
>    (8)
>    - *M* src/biome/text/cli/explore.py
>    <https://github.com/recognai/biome-text/pull/224/files#diff-9b79cd09c9dc9f1eb9a6f3f10ceee317>
>    (2)
>    - *M* src/biome/text/data/datasource.py
>    <https://github.com/recognai/biome-text/pull/224/files#diff-33c5f349688cfac08310a6e30d58a9ac>
>    (26)
>    - *M* src/biome/text/helpers.py
>    <https://github.com/recognai/biome-text/pull/224/files#diff-6a9e186184318d1a67da3a099fd38191>
>    (21)
>    - *M* src/biome/text/pipeline.py
>    <https://github.com/recognai/biome-text/pull/224/files#diff-3386de7c4987d99963cf6f544e567ad0>
>    (57)
>
> Patch Links:
>
>    - https://github.com/recognai/biome-text/pull/224.patch
>    - https://github.com/recognai/biome-text/pull/224.diff
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/recognai/biome-text/pull/224>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAIOJJY7QOFKJUZQW327MTDRTKXDJANCNFSM4NJWDN2Q>
> .
>

dcfidalgo commented on src/biome/text/data/datasource.py:
```diff
@@ -191,14 +186,12 @@ def from_yaml(
 
         if not cfg_dict.get("mapping"):
             cfg_dict["mapping"] = default_mapping or {}
-        cfg_dict.update(cfg_dict.pop("attributes", {}))
```
Personally i like the config file to reflect the init of the class, it gives you a quick way to lookup the required format of the config file.

So if we want to keep the `attribute` key in the config, i would also like to keep it in the init ... @frascuchon would that be ok for you?

frascuchon replied:
Could we just keep the original `Datasource` config and think about a better `attribute` naming?

We can do a better experience analysis of creating `DataSource`s after this PR since we need to pass them to pipeline methods

dcfidalgo replied:
@frascuchon I kept the behavior of accepting `**kwargs` in the init, but grouping them together as `reader_options` in the yaml config.
@dvsrepo @frascuchon Are you ok with the name `reader_options`?

dcfidalgo commented on src/biome/text/pipeline.py:
```diff
@@ -151,26 +154,26 @@ def train(
-        training: str,
-        validation: Optional[str] = None,
-        test: Optional[str] = None,
+        training: Union[DataSource, str],
```
done

dcfidalgo commented on src/biome/text/pipeline.py:
```diff
@@ -285,7 +294,7 @@ def load_vocabulary(self, vocab_path: str) -> None:
     def explore(
         self,
-        ds_path: str,
+        data_source: Union[DataSource, str],
```
done

dcfidalgo commented on src/biome/text/helpers.py:
```diff
@@ -99,3 +101,22 @@ def clean_metric_name(name):
+        If `data_source` is not a `DataSource` instance, just returns `data_source`
+    """
+    if isinstance(data_source, DataSource):
+        yaml_path = tempfile.NamedTemporaryFile(suffix=".yml", delete=False).name
```
just talked about it with @frascuchon , we will create the yml files in the output folder of the `train` method (see commit).
@frascuchon I just realized that for the typical flow of 1. `create_vocab` 2. `train`, we do not reuse the cached instances in the vocab creation ...

dcfidalgo commented on src/biome/text/data/datasource.py:
```diff
@@ -221,11 +214,14 @@ def to_yaml(self, path: str, make_source_path_absolute: bool = False) -> str:
 
         yaml_dict = {
             "source": source,
-            "attributes": self.attributes,
```
see comment above.

dcfidalgo commented:
Addressed the inline comments. I also moved the check for the `default_mapping` to the `Pipeline._read()` method itself, since you usually do not know this default mapping when creating a data source in the beginning. So this argument was only really used in the `Pipeline._read()` method.
